hnn_core/network.py

```diff
@@ -219,47 +286,411 @@ def __repr__(self):
+
+        drive_conn_by_cell = dict()
+        src_gid_ran_begin = self._n_gids
+
+        # Here be Dragons! It's tempting to try to refactor this to remove the
+        # duplication of the innermost for-loop, but a drive_conn must be
+        # created for every target population separately (cell_specific) or
+        # (global drive) a single drive_conn contains all target populations
+        if cell_specific:
+            for cellname in target_populations:
+                drive_conn = dict()  # NB created inside target_pop-loop
+                drive_conn['trg_gids'] = self.gid_ranges[cellname]
+                # NB list! This is used later in _parnet_connect
+                drive_conn['target_type'] = cellname
+                drive_conn['src_gids'] = range(
+                    self._n_gids, self._n_gids + len(drive_conn['trg_gids']))
+                self._n_gids += len(drive_conn['trg_gids'])
+                drive_conn['location'] = location
+
+                for receptor, weights in weights_by_receptor.items():
+                    drive_conn[receptor] = dict()
+                    if cellname in weights:
+                        drive_conn[receptor]['lamtha'] = space_constant
+                        if isinstance(dispersion_time, float):
+                            drive_conn[receptor]['A_delay'] = dispersion_time
+                        else:
+                            drive_conn[receptor][
+                                'A_delay'] = dispersion_time[cellname]
+                        drive_conn[receptor][
+                            'A_weight'] = weights[cellname]
+
+                drive_conn_by_cell[cellname] = drive_conn
+        else:
+            drive_conn = dict()
+
+            # NB list! This is used later in _parnet_connect
+            drive_conn['src_gids'] = [self._n_gids]
+            self._n_gids += 1
+            drive_conn['location'] = location
+
+            drive_conn['trg_gids'] = []  # fill in below
+            for cellname in target_populations:
+                drive_conn['trg_gids'].extend(self.gid_ranges[cellname])
+                drive_conn['target_type'] = cellname
+
+                for receptor, weights in weights_by_receptor.items():
+                    drive_conn[receptor] = dict()
+                    if cellname in weights:
+                        drive_conn[receptor]['lamtha'] = space_constant
+                        if isinstance(dispersion_time, float):
+                            drive_conn[receptor]['A_delay'] = dispersion_time
+                        else:
+                            drive_conn[receptor][
+                                'A_delay'] = dispersion_time[cellname]
+                        drive_conn[receptor][
+                            'A_weight'] = weights[cellname]
+                drive_conn_by_cell[cellname] = drive_conn
+
+        return drive_conn_by_cell, range(src_gid_ran_begin, self._n_gids)
+
+    def instantiate_drives(self, n_trials=1):
```

Comment:
make it private?

spoils the nice API page a bit: https://78-265240816-gh.circle-artifacts.com/0/html/generated/hnn_core.Network.html#hnn_core.Network

Reply:
Hmm, I agree about the spoiling. Can't decide if I agree about the privatisation though. @rythorpe opined on this earlier, too.

I'm all for letting `simulate_dipole` call the instantiation, but I do want to create an interactive example notebook for visualising the effects of choosing different parameters for the dynamics (connectivity is harder to visualise, though your idea of a conn-matrix was nice).

We could privatise now, and make public in later PR if needed.

Reply:
not sure I get your point :) what does privatization have to do with visualization?

Reply:
> what does privatization have to do with visualization?

you're right: nothing. Making private for now.

Reply:
> I'm all for letting simulate_dipole call the instantiation, but I do want to create an interactive example notebook for visualising the effects of choosing different parameters for the dynamics (connectivity is harder to visualise, though your idea of a conn-matrix was nice).

so I thought about this a bit and I think the cleanest way to achieve this is:

1. remove call to `_instantiate_drives` in dipole.py. Is this really necessary or an artifact of old code?
2. pass `n_trials` to `attach_xxx_drive`? alternatively pass it to `Network` if we want to ensure equal trials across all drives (more desirable?)
3. do self._n_trials = n_trials so user cannot modify n_trials after network is instantiated. No need for more sophisticated solution as of now :)
4. Don't call `instantiate drives` during Network creation. Instead, rename it to `_get_event_times` and call it from each of the `attach_xxx_drive` functions so you can do `self.external_drives[name]['events'] = `_get_event_times_multitrial` and this guy `_get_event_times_multitrial` would live inside `feed.py` (which should probably be renamed to input_drives.py?). It's only purpose would be to wrap `_drive_cell_event_times` with a for loop.

I am finding the nomenclature of "instantiation" a bit confusing ... the drive events should be created as soon as the user calls the method

Reply:
I see your point about exposing, but now that `_instantiate_drives` is private, perhaps OK to await future PR? This is touched by #233

Note that because `dipole.py:simulate_dipole` can be called with `n_trials=`, we need to support re-"instantiating" the event times. If anything, perhaps `simulate_dipole` should be the only place to "instantiate"?

Reply:
okay to wait for future PR. I think `n_trials=` should be removed from `simulate_dipole` and added to `Network` to support a simpler flow of code and avoid a "delayed" instantiation. Complexity grows really fast in software if we're not careful